nwapp/utils/notifier.js

```diff
@@ -28,14 +36,85 @@ function playNotificationSound() {
   audio = null;
 }
 
+
+// Cache lookup
+var urlFilePathMap = new Map();
+function getFileAtUrl(url) {
+  return new Promise(function(resolve, reject) {
+    var cachedFilePath = urlFilePathMap.get(url);
+    if(cachedFilePath) {
+      resolve(cachedFilePath);
+    }
+    else {
+      setTimeout(function() {
+        reject(new Error('Took to long to request file', url));
+      }, 750);
+
+      var tempFileStream = temp.createWriteStream();
+      tempFileStream
+        .on('finish', function() {
+          var tempFilePath = tempFileStream.path;
+          // Store it on our map as a cache lookup
+          urlFilePathMap.set(url, tempFilePath);
+          resolve(tempFilePath);
+        })
+        .on('error', function(err) {
+          reject(err);
+        });
+
+      request(url)
+        .on('error', function(err) {
+          reject(err);
+        })
+        .pipe(tempFileStream);
+    }
+  });
+}
+
 module.exports = function (spec) {
```

Comment:
I know this isnt you, but please never have a "spec" object with required properties. If an argument is is required, then it should be explicitly an argument. Optional arguments are last or in an options object, thats fine. I'm sure you know this already, but hey.

So, this function is used in three places, in three different ways:

```javascript
notifier({
  title:   msg.title,
  message: msg.text,
  roomId:  msg.troupeId,
  link:    msg.link,
  click: function () {
    navigateWindowTo(msg.link);
  }
});
```

```javascript
notifier({
  title: 'Gitter ' + version + ' Available',
  message: 'Head over to gitter.im/apps to update.'
});
```

```javascript
notifier({
  title: 'Gitter ' + version + ' Available',
  message: 'Click to restart and apply update.',
  click: function() {
    ... (does stuff)
  }
});
```

There are some required arguments and some optional ones (and some ignored ones!). Can you please change the function arguments to reflect this? Also, you may need to check through your existing code, as there are some parts that assume that an optional parameter will always be there. This is why the "spec" pattern sucks.

Also, our user_notification "msg" already includes an `icon` property that is a url. That may make your life a little easier...